### ISSUE TITLE
Example for variable table valued functions does not work.

### ISSUE THREAD

**Issue by RayWhitmer** [MicrosoftDocs/sql-docs#8781]
The code given at
https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/native-client-odbc-how-to/use-table-valued-parameters-odbc.md

has a function called demo_variable_TVP_binding.

On line 795:
 SQLPutData(hstmt, &Qty, sizeof(SQLINTEGER));

This does not capture the return value, but the next line apparently wants to test the return value.

Thus guarantees that the function will always fail without accomplishing its intended function because r is guaranteed to be SQL_NEED_DATA by the surrounding while loop.

If we fix this line by adding "r = " to be beginning of this line, the function still doesn't work, because despite the calls to tell the driver no more rows ( r = SQLPutData(hstmt, SQLPOINTER(1), 0); on line 782, the SQLParamData function on line 804 keeps returning SQL_NEED_DATA and asking for another row with the code as written here, which seems like it has not been tested. I can't make the function work.

This is the same result I got in my own coding attempts, it keeps asking for more rows despite there being no more rows. I really need an example program that works with specification of minimum driver versions that make it work (if that might be the second problem). I am currently running SQL Server Native Client 11.0 on Windows 10, since that has been successful for many other implementations. I can look for more details if necessary. I get a little scared by all the different driver version branches, e.g. native versus etc.

**Comment by RayWhitmer** [comment]
Title should probably be "Table Valued Parameters" function bindings does not work.

It has been assigned as priority 3. But it is pretty significant to me that I can't find an example showing that binding table valued parameters rows as SQL_DATA_AT_EXEC works. We need better bulk processing for SQL Server and this has derailed that effort.

Doing it as a fixed binding instead of SQL_DATA_AT_EXEC  might be a work-around, assuming those functions work better than the variable binding, but it seems even more complex and unwieldy presumably passing arrays of column values and doesn't seem to scale well to higher row counts.

Is there another example of table value parameter variable binding that works or an environment where this example is known to work (with the described fix)?

**Comment by rwestMSFT** [comment]
@RayWhitmer Thanks for your feedback.

The priority rating of the issue is automated, as no one has triaged this issue yet.

That being said, I wonder if this is better suited as a support request. If it's a documentation issue, by all means we should fix that, but if you think this is a shortcoming in the product, there's not much we can do except suggest opening a support case.

One of my colleagues noted that you're using SNAC 11 on Windows 10. SNAC was [removed](https://learn.microsoft.com/en-us/archive/blogs/sqlreleaseservices/snac-lifecycle-explained) from SQL Server 2022, and hasn't been serviced for a number of years, so we have two possible scenarios, both of which suggest this becoming a support issue:

- If there is no requirement for SNAC over OLEDB, please try SNAC 13, but we would need a repro with the latest OLEDB or ODBC drivers.
- If you require SNAC over OLEDB support, the [SQL Server 2012 SNAC driver](https://www.microsoft.com/en-us/download/details.aspx?id=50402) was the last one that supported that functionality.

In either case, a repro of the issue using the latest OLEDB / ODBC drivers would help us fix the article in question.

**Comment by RayWhitmer** [comment]
I will need fixed documentation that reflects the actual functionality to determine whether it might be a product shortcoming. If the fix to the documentation is to change it to say, in all the places that describe this function, you can't successfully perform this feature, then it is a feature shortcoming and I can take it up there. If the fix is to say, that really should work, then I would expect you to at a minimum endorse a bug report, if the basic sample doesn't work.

I can try to do it with a more up-to-date driver, since that seems like a prerequisite and I don't know where I would find a list of fixed bugs. Who knows, but I would have expected a search to turn something up if it were something known to have been broken for an extended period of time. As a resolution I would at least expect it to show up in the documentation if it was broken in significant versions of the driver.

I assume SNAC means native driver, so I will double check with native driver 13. I don't think we use OLEDB, but I am not sure.

**Comment by RayWhitmer** [comment]
I misread. I will try the latest ODBC driver. I think last time I tried it, it failed a lot for our products, seemed like compatibility wasn't there, but I will try again.

**Comment by RayWhitmer** [comment]
I just redefined the sources with ODBC driver 17. No failures at all in our product. This produced exactly the same behavior, an unterminated loop of SQL_NEED_DATA. I will recommend upgrading all our environments to ODBC driver 17. But that apparently has nothing to do with the problem. It could be that your example is using the API incorrectly. Or it is broken with respect to the sample code.

**Comment by RayWhitmer** [comment]
How about I ask you, who said this sample worked when they gave it to you? It was obviously broken. How can you have sample code that no one stands behind? Give me the details. Who called these sample functions and verified that they worked? Can you get them to give you a version that at least works for them in their environment?

**Comment by rwestMSFT** [comment]
@RayWhitmer The contents of this article haven't changed since it was migrated to Database Docs in 2017, and could easily be decades old. As it stands right now, we don't have enough information to know if this is a bug, or that the documentation is incorrect.

If you have a repro of the issue, then your next step is to open a support case. Once that's done, the developers will investigate and decide whether it's a bug. Once that's resolved, we will be directed to fix the code sample as appropriate.

**Comment by rwestMSFT** [comment]
The code sample has been around for over 15 years according to the MSDN archives I found. I have let the product group know that the code needs review, and it may take weeks or months to update, given the size of their backlog.

In the meantime, I did some digging and found these links that might be useful, although I expect you've already found them:

- https://learn.microsoft.com/sql/odbc/reference/syntax/sqlputdata-function
- https://learn.microsoft.com/sql/odbc/reference/syntax/sqlbindparameter-function
- https://learn.microsoft.com/sql/relational-databases/native-client-odbc-table-valued-parameters/sending-data-as-a-table-valued-parameter-using-data-at-execution-odbc
- https://stackoverflow.com/questions/84064/sqlbindparameter-to-prepare-for-sqlputdata-using-c-and-sql-native-client

If these are helpful, and you're able to figure out the problem, you would help the Database Docs team greatly by submitting a pull request to update the article sooner than we'll be able to get to it.

**Comment by RayWhitmer** [comment]
**Please give me the link to report this to support.** I have reported multiple MSVC-related issues in the past, but I don't remember where and the places I found using a search don't look appropriate for reporting ODBC binding issues.

Thanks for the links. I have seen these and tried variations. The sources don't seem consistent with each other in various ways. And I doubt any of them work.

For example, the above source ([sending-data-as-a-table-valued-parameter-using-data-at-execution-odbc](sending-data-as-a-table-valued-parameter-using-data-at-execution-odbc) says:

    0,         // DecimalDigits: For a table-valued parameter this is always 0.

Yet another source for the same use case (the one we have been discussing at [use-table-valued-parameters-odbc.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/native-client-odbc-how-to/use-table-valued-parameters-odbc.md)) says:

    0,// DecimalDigits - for a TVP this is the number of columns in the TVP

They are both passing 0, but one says it should be the number of columns.

And the prior argument that is called ColumnSize which should be the number of rows according to comments I think is just an internal buffer size, not the number of rows, as the examples claim, and in the examples, it never matches the number of rows being sent, and there is a lot more ambiguity in addition to outright broken code.

**Comment by rwestMSFT** [comment]
I discussed your issue at length with a PM in the product group, and given that we've removed SNAC from SQL Server 2022, this article falls within the parameters of "needs updating" *anyway*, because it references `sqlncli.h`.

This, along with your original and well-stated concerns, should escalate it internally. That means a days-to-weeks timeframe, as opposed to weeks-to-months.

[This page](https://learn.microsoft.com/sql/sql-server/sql-server-get-help) outlines various methods of opening a support case, depending on license agreements you have with Microsoft.

Thank you for bearing with me as I got to grips with the issue. It's one of the risks of decades-old products, that we have decades-old content to match that hasn't been reviewed until someone like yourself asks questions.

**Comment by RayWhitmer** [comment]
We are a smaller company inside a much larger one. I have been trying to get the larger one to cough up information required for filing a Microsoft ticket. I don't know if other parts of the company do serious back-end development on Windows. It is hard to even figure out what "product" the ticket would be logged under. I am still trying to make progress and have not given up.

The bigger picture is that I have created a C++ variadic binding to ODBC to make it trivial to use the database, just pass the statement and the parameters in a single call, and the C++ templates figure out all the ODBC binding and hide it from the user. In cases of bulk input/output a callback is used that has arguments corresponding to the columns involved, still much easier than ODBC bindings, which are automated underneath the covers based on parameter types. If we could make table valued parameters work for us, it would soon be heavily used because we have lots of use cases and it would become easy once supported by our C++ variadic binding. Microsoft ought to do something like a variadic much-simplified binding in C++ (O don't think OLE counts, but I don't know).

While it may be less efficient, I am thinking of re-implementing bulk parameters as a JSON string that is parsed by the stored procedure. As time ticks by, the case becomes much stronger to just dump the code supporting table valued parameters. There are other reasons for this as well, e.g. portability to Linux where it should be easier to resolve issues such as this one, and other development issues encountered by other groups with ODBC to target SQL Server on Linux (e.g. interfering with optimal versions of cryptographic libraries that products need). I have run Postgres with 50TB of quite active data for millions of users and have no complaints and retargeting the variadic binding at Postgres is very doable.

But if we had even a sample of variable table valued parameter bindings that worked in SQL Server, we might be able to use the code I have written for the variadic binding as-is, at least in the short term and maybe the longer term -- there were reasons we decided to use it. This is a frustrating spot to be in because my work to make TVP work seems wasted thus far.

Thanks for your attention and help thus far.

**Comment by RayWhitmer** [comment]
No success getting the powers that be to agree to give me info to open a support case.

I have thought of a lot of different variations that I have tried with no success.

I just tried making the basic table binding with SQL_LEN_DATA_AT_EXEC(row_count) instead of SQL_DATA_AT_EXEC, but this makes SQLParamData return SQL_ERROR (-1) after returning returning SQL_NEED_DATA once. No longer an unterminated loop, but still a failure: Failed database operation
+{"_id":"SQLDiag","code":65535}
+{"_id":"SQLDiagRec","code":0,"message":"[Microsoft][ODBC Driver 17 for SQL Server]String data, length mismatch","state":22026}

If I try returning, say, 100 rows by calling SQLPutData 100 times before calling SQLParamData, then SQLParamData no longer returns the error, but it goes back to the unterminated loop behavior.

We are headed towards a JSON solution, but I don't think it can perform as well.

**Comment by RayWhitmer** [comment]
(WRT the "String data length error, I wasn't binding a string anywhere in the table.)

**Comment by RayWhitmer** [comment]
My argument to my superiors is if we can't get them to pay for "expensive" support for MS products, then the MS products are effectively out of support.

**Comment by rwestMSFT** [comment]
@RayWhitmer I'm sorry to hear that. For what it's worth, any costs you incur for a support case is usually refunded if there is a bug.

**Comment by RayWhitmer** [comment]
I got authorization to log an issue with Microsoft, with a credit card, but it has been too difficult to figure out how to do so. Starting at the phone number 800-642-7676, the automated menus are completely stupicd and don't understand, e.g. "broken odbc example". The whole system is apparently designed to discourage people from raising issues. The best I get is a recommendation to log an issue at aka.ms, which immediately tells me "You don't have access to this".

**Comment by rwestMSFT** [comment]
Ray, much as I understand your frustration, there's nothing more I can do. I was however able to find this link:

https://support.serviceshub.microsoft.com/supportforbusiness/create

Good luck.

**Comment by RayWhitmer** [comment]
That would have been a nice link to get initially. The phone number (800) 642 7676 directs most things to aka.ms, which is fairly worthless.

Using your new link, I have logged a ticket. We will hope for something good.

**Comment by rwestMSFT** [comment]
Respectfully, that was the first link on the [original URL](https://learn.microsoft.com/sql/sql-server/sql-server-get-help) I sent you in March.

![CleanShot 2023-06-20 at 14 12 30@2x](https://github.com/MicrosoftDocs/sql-docs/assets/97149825/3b47077a-9c2b-42ae-afa8-61ebf2ad3088)

I am closing this issue for now. When we have some feedback from the product group to update this article, I will let you know.

**Comment by RayWhitmer** [comment]
Respectfully it is not. I do not see how to get from https://user-images.githubusercontent.com/97149825/247249071-3b47077a-9c2b-42ae-afa8-61ebf2ad3088.png from https://support.serviceshub.microsoft.com/supportforbusiness/create and just because you think the various URLs are interchangeable doesn't mean they are in practice.

**Comment by RayWhitmer** [comment]
And now support is closing the issue without even investigating it. SQL Server is the wrong category, they say, but no one will tell me what the right category is.

No one will own the functionality or the documentation or the fact that it is wrong.

Please give me your name and the email of your supervisors.

**Comment by RayWhitmer** [comment]
The support call I paid $499 for at the URL you gave me goes to a non-Microsoft employee in India, who have no access to anyone they would need to troubleshoot a problem.

**Comment by RayWhitmer** [comment]
SR 2306200030004941

I sent a lot of emails to managers up the tree and finally got a 10-minute meeting with the person who was assigned and presumably some of the others I emailed. I do not know the actual status of the ticket because I have no way to view that that I am aware of.

They asked me to show them the problem. I started stepping them through the debugger pointing out that this was Microsoft sample code, and they immediately lost interest and didn't want me to even tell them what was happening or should have happened. They said they do not deal with code, and they are not aware of any support group that deals with code.

(Microsoft must be really dumbed down, most organizations require coding from various levels of support, at least when I was hiring them, but maybe India is different).

I noticed in the support ticket legalese was a disclaimer stating that the ticket must not be about custom code, although arguably this is not custom code, this is Microsoft sample code. If it is a problem, then it is a problem that you told me to go to support when there is no support organization willing to talk about e.g. ODBC Table Valued Function APIs.

They said they would be evaluating things and get back to me within a few days.
